src/boxed.rs

```diff
@@ -0,0 +1,208 @@
+//! A pointer type for heap allocation.
+
+use crate::alloc;
+use std::{borrow, fmt, hash, mem, ptr};
+
+/// A loom version of `std::boxed::Box` based on the leak tracking in `loom::alloc`.
+///
+/// # Known problems
+///
+/// This `Box` version can't be dereferenced in order to take the value
+/// from the heap and bring it back to the stack. This is because `std::boxed::Box`
+/// is tightly integrated with the compiler and uses magic that normal crates can't.
+/// This version instead provides [`Box::into_value`] which does the same thing.
+pub struct Box<T: ?Sized> {
+    ptr: *mut T,
```

Comment:
I can't comment on the issue of whether this is a good idea, but this should be a `*const T` to give the box the correct [variance][1].

[1]: https://doc.rust-lang.org/stable/nomicon/subtyping.html

Reply:
That is great input. I added tests to make sure the size and alignment of `Box<T>` and `Option<Box<T>>` stays the same as the standard library box for a few different `T`s. The tests failed. But they passed after changing to `NonNull`.

Fixed!